metadata-ingestion/src/datahub/ingestion/source_config/operation_config.py

```diff
@@ -0,0 +1,86 @@
+import datetime
+import logging
+from typing import Any, Dict, Optional
+
+import pydantic
+from pydantic.fields import Field
+
+from datahub.configuration.common import ConfigModel, ConfigurationError
+
+logger = logging.getLogger(__name__)
+
+
+class OperationConfig(ConfigModel):
+    lower_freq_profile_enabled: bool = Field(
+        default=False,
+        description="Whether to do profiling at lower freq or not. This does not do any scheduling just adds additional checks to when not to run profiling.",
+    )
+    profile_day_of_week: Optional[int] = Field(
+        default=None,
+        description="Number between 0 to 6 for day of week (both inclusive). 0 is Monday and 6 is Sunday. If not specified, defaults to Nothing and this field does not take affect.",
+    )
+    profile_date_of_month: Optional[int] = Field(
+        default=None,
+        description="Number between 1 to 31 for date of month (both inclusive). If not specified, defaults to Nothing and this field does not take affect.",
+    )
+
+    @pydantic.root_validator(pre=True)
+    def lower_freq_configs_are_set(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        lower_freq_profile_enabled = values.get("lower_freq_profile_enabled")
+        profile_day_of_week = values.get("profile_day_of_week")
+        profile_date_of_month = values.get("profile_date_of_month")
+        if (
+            lower_freq_profile_enabled
+            and profile_day_of_week is None
+            and profile_date_of_month is None
+        ):
+            raise ConfigurationError(
+                "Lower freq profiling setting is enabled but no day of week or date of month is specified. Profiling will be done.",
+            )
+        return values
+
+    @pydantic.validator("profile_day_of_week")
+    def validate_profile_day_of_week(cls, v: Optional[int]) -> Optional[int]:
+        profile_day_of_week = v
+        if profile_day_of_week is None:
+            return None
+        if profile_day_of_week < 0 or profile_day_of_week > 6:
+            raise ConfigurationError(
+                f"Invalid value {profile_day_of_week} for profile_day_of_week. Must be between 0 to 6 (both inclusive)."
+            )
+        return profile_day_of_week
+
+    @pydantic.validator("profile_date_of_month")
+    def validate_profile_date_of_month(cls, v: Optional[int]) -> Optional[int]:
+        profile_date_of_month = v
+        if profile_date_of_month is None:
+            return None
+        if profile_date_of_month < 1 or profile_date_of_month > 31:
+            raise ConfigurationError(
+                f"Invalid value {profile_date_of_month} for profile_date_of_month. Must be between 1 to 31 (both inclusive)."
+            )
+        return profile_date_of_month
+
+
+def is_profiling_enabled(operation_config: OperationConfig) -> bool:
```

Comment:
This PR just got my attention because this feature makes totally sense: I want to decouple the scheduling of the profiling from the _basic_ ingestion without having to duplicate the recipes. So glad to see this!

About this `is_profiling_enabled` implementation, let me test my understanding of the code. If I run an ingestion on hourly basis and I want the profiling to be executed _eg_ once in a month `profile_date_of_month=1`, this will result on the profiling being executed on all the hours for the day 1, in this example. Am I wrong? Unless I'm wrong with my analysis, I would say this is not the desired behavior. 

Solving this will likely require to save some state to determine if the profiling has been already executed.
Alternatively, the issue may also get solved by giving more _precision_ to the profiling scheduling beyond weekday XOR monthday. Instead, full cron expression may be set and then check if it is the time for running profiling. 

This is just a library supporting this case https://github.com/kiorky/croniter It provides functionality to [test whether a date matches cron expression](https://github.com/kiorky/croniter#testing-if-a-date-matches-a-crontab); rounding the date to _eg_ the hour may be enough and mitigate the need to _save state_.

Reply:
We're thinking about this PR as a stop-gap solution to fix the pain point, but know that it's not going to handle all use cases well and is a bit bespoke. Eventually, we want to move towards something that looks more like this:

```yaml
profiling.enabled:
  - operator: day_of_week_matches
    values: ["SUN"]
  - operator: day_of_month_matches
    values: ["5"]
```

We'd have a broader set of operators + ability to use custom user-provided operators, which should afford us the flexibility that we need.

But definitely looking for thoughts and feedback on this, as well as other use cases that you're thinking about.

Reply:
@sgomezvillamor This is a stop gap measure. 

The specific case this tries to solve for now is to for the very large deployments where getting schema is very fast but doing profiling is very costly because of size of tables being in 100s of GBs or TB scale. In those cases we are seeing people either running the recipe once a week or maintaining 100s of recipes as duplicates. A recent PR was done https://github.com/datahub-project/datahub/pull/8317 for supporting those duplication cases where someone wanted to maintain the recipes in their code with and without profiling and setting them on different schedules to work-around this limitation.

At the end of the day we need improvements in either the scheduler that runs in server side or more operators in recipes that help with more cases on the ingestion side. But that is a larger lift. We can always improve connector specific profiling to make sure it stores the specific information for that ingestion recipe but that involves lot more effort.

We will definitely iterate on this based on feedback that we see with orgs as we see the pain points.

Reply:
Now this is abstracted away into a class we can easily add a `enabled_v2` and add a new set of flags (or use existing flags into a new logic) and the method inside it will take care of adding the functionality to all connectors across the board. So it becomes very easy to iterate and change the functionality without worrying about breaking anything.

Reply:
But you may want to look at https://github.com/datahub-project/datahub/pull/8317 which I mentioned as at least 1 large org has found success using that to maintain multiple recipes in their code and just writing small script to maintain the recipes. 

Recipes-as-code they called it. While not as elegant as something like terraform it does help with maintaining a large number of recipes.

Reply:
> The specific case this tries to solve for now is to for the very large deployments where getting schema is very fast but doing profiling is very costly because of size of tables being in 100s of GBs or TB scale. In those cases we are seeing people either running the recipe once a week or maintaining 100s of recipes as duplicates. 

Yes, my team is one of those 😅 

My current scenario is my org usually runs connectors on hourly-basis, while we run profiling on weekly/monthly basis. With current implementation in this PR we may set _eg_ `profile_date_of_month=1` and that will result on being skipped for all days in the month but day 1. However on day 1 profiling is executed 24 times. Not ideal at all.

Since granularity of the profiling scheduling is days (week day or month day), current implementation is limited to schedulings based on days, weeks, months... not hours. I was just pointing out this in case you missed it. Glad to hear that this is just the starting point of a promising feature and hopefully scenarios such as mine are also addressed in the future 💪 Thanks for the detailed responses, really appreciated.